File: problems_claude/1.py
# Problem
# **Difficulty: ** Easy
# **Time: ** 15 minutes
# **Pattern: ** Array traversal with edge cases
#
# Given an array of integers, create a new array where each element is the sum of itself and its immediate neighbors.
# Handle edge cases for first and last elements.
#
# ```python
#
#
def neighbor_sum(arr: list[int]) -> list[int]:
    """
    Create an array where each element is the sum of itself and its neighbors.

    Examples:
    Input: [1, 2, 3, 4, 5]
    Output: [3, 6, 9, 12, 9]

    Explanation:
    - arr[0]: 1 + 2 = 3 (no left neighbor)
    - arr[1]: 1 + 2 + 3 = 6
    - arr[2]: 2 + 3 + 4 = 9
    - arr[3]: 3 + 4 + 5 = 12
    - arr[4]: 4 + 5 = 9 (no right neighbor)
    """
    # Your code here
    return_array = []
    for i, value in enumerate(arr):
        if i == 0 and len(arr) == 1:
            return_array = arr
        elif i == 0:
            return_array.append(value + arr[i + 1])
        elif i == len(arr) - 1:
            return_array.append(value + arr[i - 1])
        else:
            return_array.append(arr[i - 1] + value + arr[i + 1])
    return return_array


# Test cases
assert neighbor_sum([1, 2, 3, 4, 5]) == [3, 6, 9, 12, 9]
assert neighbor_sum([10]) == [10]
assert neighbor_sum([5, 1]) == [6, 6]
assert neighbor_sum([]) == []
print("All tests passed!")

#
#
#
#
#
#

File: cpp/problems_1/1.cpp
// #### Problem 1: Two Sum
// **Difficulty:** Easy
// **Time:** 15 minutes
// **Pattern:** Hash map for complement lookup
//
// Given an array of integers `nums` and an integer `target`, return indices of
// the two numbers such that they add up to `target`.

#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

std::vector<int> twoSum(std::vector<int> &nums, int target) {
  /*
  Find two numbers that add up to target and return their indices.

  Examples:
  Input: nums = [2,7,11,15], target = 9
  Output: [0,1] (nums[0] + nums[1] = 2 + 7 = 9)

  Input: nums = [3,2,4], target = 6
  Output: [1,2] (nums[1] + nums[2] = 2 + 4 = 6)
  */

  // Your code here
  for (auto &e : nums) {
    auto to_find = target - e;
    std::cout << "To find " << to_find << std::endl;
  }
  return {};
}

int main() {
  std::vector<int> nums1 = {2, 7, 11, 15};
  std::vector<int> result1 = twoSum(nums1, 9);
  assert(result1.size() == 2 && result1[0] == 0 && result1[1] == 1);

  std::vector<int> nums2 = {3, 2, 4};
  std::vector<int> result2 = twoSum(nums2, 6);
  assert(result2.size() == 2 && result2[0] == 1 && result2[1] == 2);
  //
  std::cout << "Problem 1: All tests passed!" << std::endl;
  return 0;
}
